### ISSUE TITLE
[BUG] Missing syncthreads in generated kernel

### ISSUE THREAD

**Issue by hyx1999** [tile-ai/tilelang#1822]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

0.1.7.post3

### System information

3.12.12 (main, Dec  9 2025, 19:02:36) [Clang 21.1.4 ] linux
0.1.7.post3
2.7.1+cu126

### Problem description

In the following code, if sync_threads is not manually added, it will result in an incorrect Mask

without sync_threads:
Mask:
 tensor([[0, 0, 0, 1, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0],
        [1, 0, 0, 0, 0, 0, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1]], device='cuda:0', dtype=torch.int32)

with sync_threads:
Mask:
 tensor([[0, 0, 0, 1, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0],
        [1, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 1, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 1, 1, 1, 1],
        [0, 0, 0, 0, 1, 1, 1, 1],
        [0, 0, 0, 0, 1, 1, 1, 1]], device='cuda:0', dtype=torch.int32)

### Reproducible example code

The Python snippets:

```python
import torch
import tilelang as tl
import tilelang.language as T
from typing import Optional
from einops import rearrange, repeat


@tl.jit(pass_configs={
    tl.PassConfigKey.TL_DISABLE_TMA_LOWER: True,
    tl.PassConfigKey.TL_DISABLE_WARP_SPECIALIZED: True,
    tl.PassConfigKey.TL_DISABLE_DATA_RACE_CHECK: True,
})
def mha_update_block_indices_kernel(
    topk: int,
    max_blocks: int,
    num_kv_heads: int,
    threads: int = 256,
):
    batch = T.symbolic("batch")

    block_indices_shape = [batch, num_kv_heads, topk]
    load_shape = [batch, num_kv_heads]

    @T.prim_func
    def main(
        PreBlockIndices: T.Tensor(block_indices_shape, T.int32),
        CurBlockIndices: T.Tensor(block_indices_shape, T.int32),
        NextBlockIndices: T.Tensor(block_indices_shape, T.int32),
        SrcIndices: T.Tensor(block_indices_shape, T.int32),
        DstOffsets: T.Tensor(block_indices_shape, T.int32),
        LoadBlocks: T.Tensor(load_shape, T.int32),
        Debug: T.Tensor([2, topk], T.int32),
        Mask: T.Tensor([topk, topk], T.int32),
    ):
        with T.Kernel(batch, num_kv_heads, threads=threads) as (bx, by):
            i_b = bx
            i_h = by
            pre_indices_shared = T.alloc_shared([topk], T.int32)
            cur_indices_shared = T.alloc_shared([topk], T.int32)
            nxt_indices_shared = T.alloc_shared([topk], T.int32)
            mask_shared = T.alloc_shared([topk, topk], T.int32)
            mask = T.alloc_fragment([topk, topk], T.int32)
            T.copy(PreBlockIndices[i_b, i_h, :], pre_indices_shared)
            T.copy(CurBlockIndices[i_b, i_h, :], cur_indices_shared)
            T.copy(pre_indices_shared, nxt_indices_shared)

            # T.sync_threads() # BUG: without T.sync_threads() cause bug
            T.fill(mask_shared, 0)
            for i, j in T.Parallel(topk, topk):
                mask_shared[i, j] = T.if_then_else(
                    cur_indices_shared[i] == pre_indices_shared[j],
                    1, 0
                )
            T.copy(mask_shared, mask)
            for i in T.Parallel(topk):
                Debug[0, i] = cur_indices_shared[i]
                Debug[1, i] = pre_indices_shared[i]
            for i, j in T.Parallel(topk, topk):
                Mask[i, j] = mask_shared[i, j]


    return main

class KVO_MHAStoreKVInterface:

    def __init__(self,
        max_batch: int,
        max_seqlen: int,
        num_kv_heads: int,
        dim_k: int,
        dim_v: int,
        topk: int,
        block_size: int,
    ) -> None:
        max_blocks = max_seqlen // block_size
        self.topk = topk
        self.block_size = block_size
        self.num_kv_heads = num_kv_heads
        self.max_blocks = max_seqlen // block_size

        self.update_block_indices_kernel = mha_update_block_indices_kernel(
            topk,
            max_blocks,
            num_kv_heads,
        )

    def update_block_indices(
        self,
        pre_block_indices: torch.Tensor,
        cur_block_indices: torch.Tensor,
    ):
        device = cur_block_indices.device
        load_blocks = torch.zeros(pre_block_indices.shape[:2], dtype=torch.int32, device=device)
        next_block_indices = torch.zeros_like(pre_block_indices)
        src_indices = torch.zeros_like(pre_block_indices)
        dst_offsets = torch.zeros_like(pre_block_indices)
        Debug = torch.zeros((2, self.topk), dtype=torch.int32, device=device)
        Mask = torch.zeros((self.topk, self.topk), dtype=torch.int32, device=device)
        self.update_block_indices_kernel(
            pre_block_indices,
            cur_block_indices,
            next_block_indices,
            src_indices,
            dst_offsets,
            load_blocks,
            Debug,
            Mask,
        )
        print("pre_block_indices:\n", pre_block_indices)
        print("cur_block_indices:\n", cur_block_indices)
        print("Debug:\n", Debug[:, :8])
        print("Mask:\n", Mask[:16, :16])
        return next_block_indices, src_indices, dst_offsets, load_blocks


if __name__ == "__main__":
    import torch
    import numpy as np
 
    def test_update_block_indices():
        print("Testing update_block_indices...")
        
        # 设置参数
        max_batch = 4
        max_seqlen = 1024
        num_kv_heads = 1
        dim_k = 64
        dim_v = 64
        topk = 8
        block_size = 128
        
        # 创建接口实例
        interface = KVO_MHAStoreKVInterface(
            max_batch=max_batch,
            max_seqlen=max_seqlen,
            num_kv_heads=num_kv_heads,
            dim_k=dim_k,
            dim_v=dim_v,
            topk=topk,
            block_size=block_size
        )
        
        # 创建测试数据
        batch_size = 1
        
        # 前一个时间步的block indices
        pre_block_indices = torch.tensor(
            [1, 3, 5, 7, -1, -1, -1, -1] + [-1] * (topk - 8), 
            dtype=torch.int32
        ).cuda().view(1, 1, -1)
        
        # 当前时间步的block indices（模拟一些变化）
        cur_block_indices = torch.tensor(
            [7, 4, 6, 1, 5, -1, -1, -1] + [-1] * (topk - 8), 
            dtype=torch.int32
        ).cuda().view(1, 1, -1)
                
        # 执行更新
        next_block_indices, src_indices, dst_offsets, load_blocks = interface.update_block_indices(
            pre_block_indices, cur_block_indices
        )
        
    test_update_block_indices()

```

### Traceback

```pytb

```

**Comment by LeiWang1999** [comment]
resolved with pr #1829 , thanks for your report!
